diff --git a/MAINTAINERS b/MAINTAINERS
index b22e7fd..969e625 100644
--- a/MAINTAINERS
+++ b/MAINTAINERS
@@ -7630,6 +7630,11 @@ M:	Juanjo Ciarlante <jjciarla@raiz.uncu.edu.ar>
 S:	Maintained
 F:	net/ipv4/netfilter/ipt_MASQUERADE.c
 
+SAIL FIB LOOKUP
+M:	MD Iftakharul Islam (Tamim) <tamim@csebuet.org>
+S:	Maintained
+F:	net/ipv4/fib_sail_xdp.c
+
 IPMI SUBSYSTEM
 M:	Corey Minyard <minyard@acm.org>
 L:	openipmi-developer@lists.sourceforge.net (moderated for non-subscribers)
diff --git a/include/net/ip6_fib.h b/include/net/ip6_fib.h
index 3d49305..483a8ce 100644
--- a/include/net/ip6_fib.h
+++ b/include/net/ip6_fib.h
@@ -40,6 +40,97 @@
 struct rt6_info;
 struct fib6_info;
 
+#if IS_ENABLED(CONFIG_FIB_SAIL_XDP)
+/*
+ * The router can have upto 255 ports. This limitation
+ * allows us to represent netdev_index as an u8
+ */
+#define NETDEV_COUNT_MAX 255
+#define PARTS 1024
+
+/*Chunk of size 1024*/
+struct chunk16 {
+	/*256-bit bitmap. Here i-th bit (from LSB) is set to 1 if C24[i] > 0 */
+	u64 c_bitmap[PARTS];
+	/*
+	 * Index to C24 where chunk is started. A chunk corresponds
+	 * to 256 elements. Instead of having just one start index for the
+	 * whole chunk, we divide the chunk into 4 parts and save start
+	 * index for each part.
+	 */
+	u16 c_popcnt[PARTS];
+        
+	/*256-bit bitmap. Here i-th bit (from LSB) is set to 1 if C24[i] > 0 */
+	u64 n_bitmap[PARTS];
+	/*
+	 * Index to C24 where chunk is started. A chunk corresponds
+	 * to 256 elements. Instead of having just one start index for the
+	 * whole chunk, we divide the chunk into 4 parts and save start
+	 * index for each part.
+	 */
+	u16 n_popcnt[PARTS];
+};
+
+#define SMALL 4
+/*Chunk of size 4*/
+struct chunk8 {
+	/*256-bit bitmap. Here i-th bit (from LSB) is set to 1 if C24[i] > 0 */
+	u64 bitmap[SMALL];
+	/*
+	 * Index to C24 where chunk is started. A chunk corresponds
+	 * to 256 elements. Instead of having just one start index for the
+	 * whole chunk, we divide the chunk into 4 parts and save start
+	 * index for each part.
+	 */
+	u16 popcnt[SMALL];
+};
+
+struct sail {
+	/*default next-hop (Level 0)*/
+	u8	def_nh;
+
+	/*Level 16*/
+	u8 __rcu *N16;
+	u8 __rcu *P16;
+	u16 __rcu *C16;
+        
+	/*Level 32*/
+	u8 __rcu *N32;
+	u8 __rcu *P32;
+	struct chunk16 __rcu *CK32;
+	u32 cnk32_count;/*Number of chunks in level 32*/
+        u64 n32_cnt;/*Number of elements in N32*/
+
+
+	/*Level 40*/
+	u8 __rcu *N40;
+	u8 __rcu *P40;
+	struct chunk8 __rcu *CK40;
+	u32 cnk40_count;/*Number of chunks in level 40*/
+        
+	/*Level 48*/
+	u8 __rcu *N48;
+	u8 __rcu *P48;
+	struct chunk8 __rcu *CK48;
+	u32 cnk48_count;/*Number of chunks in level 48*/
+
+	/*Level 64*/
+	u8 __rcu *N64;
+	u8 __rcu *P64;
+        u32 cnk64_count;/*Number of chunks in level 48*/
+
+	/*Index to this array is stored in N16, N24 and N32*/
+	struct net_device	*netdevs[NETDEV_COUNT_MAX];
+	u8 netdev_count;/*Number of netdevs*/
+
+	spinlock_t lock;
+};
+
+int sail_insert(struct sail *s, u64 key, u8 prefix_len, struct net_device *dev);
+int sail_lookup(const struct sail *s, const u64 dest, struct net_device **dev);
+#endif
+
+
 struct fib6_config {
 	u32		fc_table;
 	u32		fc_metric;
@@ -342,6 +433,10 @@ struct fib6_table {
 	struct fib6_node	tb6_root;
 	struct inet_peer_base	tb6_peers;
 	unsigned int		flags;
+#if IS_ENABLED(CONFIG_FIB_SAIL_XDP)
+	/*Each FIB table will have its own SAIL structure.*/
+	struct sail	sail;
+#endif
 	unsigned int		fib_seq;
 #define RT6_TABLE_HAS_DFLT_ROUTER	BIT(0)
 };
diff --git a/net/ipv4/Kconfig b/net/ipv4/Kconfig
index 32cae39..5c41071 100644
--- a/net/ipv4/Kconfig
+++ b/net/ipv4/Kconfig
@@ -52,6 +52,17 @@ config IP_ADVANCED_ROUTER
 
 	  If unsure, say N here.
 
+config FIB_SAIL_XDP
+	bool "SAIL based FIB lookup for XDP"
+	depends on !IP_ROUTE_MULTIPATH && !LWTUNNEL
+	default y
+	help
+	  This option will enable SAIL based routing table lookup for XDP
+	  forwarding. This will store FIB table as a SAIL (along with
+	  LC-trie). Currently it only supports up to 255 ports. Currently
+	  it does not support multi-path routing and light-weight
+	  tunnels such as MPLS.
+
 config IP_FIB_TRIE_STATS
 	bool "FIB TRIE statistics"
 	depends on IP_ADVANCED_ROUTER
diff --git a/net/ipv6/Makefile b/net/ipv6/Makefile
index e0026fa..3cf2572 100644
--- a/net/ipv6/Makefile
+++ b/net/ipv6/Makefile
@@ -48,6 +48,7 @@ obj-$(CONFIG_IPV6_SIT) += sit.o
 obj-$(CONFIG_IPV6_TUNNEL) += ip6_tunnel.o
 obj-$(CONFIG_IPV6_GRE) += ip6_gre.o
 obj-$(CONFIG_IPV6_FOU) += fou6.o
+obj-$(CONFIG_FIB_SAIL_XDP) += fib_sail_xdp.o
 
 obj-y += addrconf_core.o exthdrs_core.o ip6_checksum.o ip6_icmp.o
 obj-$(CONFIG_INET) += output_core.o protocol.o $(ipv6-offload)
diff --git a/net/ipv6/fib_sail_xdp.c b/net/ipv6/fib_sail_xdp.c
new file mode 100644
index 0000000..72e570a
--- /dev/null
+++ b/net/ipv6/fib_sail_xdp.c
@@ -0,0 +1,851 @@
+// SPDX-License-Identifier: GPL-2.0
+/* Copyright (c) 2018-19 MD Iftakharul Islam (Tamim) <mislam4@kent.edu>
+ *
+ *   This program is free software; you can redistribute it and/or
+ *   modify it under the terms of the GNU General Public License
+ *   as published by the Free Software Foundation; either version
+ *   2 of the License, or (at your option) any later version.
+ */
+
+#include <net/ip6_fib.h>
+
+/*The length of N16, P16 and C16 is 2^16*/
+#define LEVEL16_SIZE 65536
+
+/*Total number of chunks preallocated for level 24-64*/
+#define NUM_CHUNKS 64
+
+/*Number of small chunks*/
+#define SM_CHUNKS 16384
+
+/*chunk size is 2^16*/
+#define CNK_16 65536
+
+/*chunk size is 2^8*/
+#define CNK_8 256
+
+/*memory preallocated for N and P array in level 24-64*/
+#define ARR_SIZE NUM_CHUNKS * CNK_16
+
+#define ARR_SIZE_SM SM_CHUNKS * CNK_8
+
+/*Calculates the number of bits set to 1*/
+#define POPCNT(X) (hweight64(X))
+
+/*POPCNT of left-most N bits of X*/
+#define POPCNT_OFF(X, N) (hweight64(((1ULL << (N)) - 1) & (X)))
+
+/*Calculate index to C24 from CK26 chunk and chunk offset */
+static u16 calc_ck_idx(struct chunk16 *c, u64 cnk_off)
+{
+	u32 part_idx, part_off;
+
+	part_idx = cnk_off / 64;
+	part_off = cnk_off % 64;
+
+	return c->c_popcnt[part_idx] +
+		POPCNT_OFF(c->c_bitmap[part_idx], part_off);
+}
+
+/*Calculate index to N from CK26 chunk and chunk offset */
+static u16 calc_n_idx(struct chunk16 *c, u64 cnk_off)
+{
+	u32 part_idx, part_off;
+
+	part_idx = cnk_off / 64;
+	part_off = cnk_off % 64;
+
+	return c->n_popcnt[part_idx] +
+		POPCNT_OFF(c->n_bitmap[part_idx], part_off);
+}
+
+static u16 calc_ck_idx_sm(struct chunk8 *c, u64 cnk_off)
+{
+	u8 part_idx, part_off;
+
+	part_idx = cnk_off / 64;
+	part_off = cnk_off % 64;
+
+	return c->popcnt[part_idx] +
+		POPCNT_OFF(c->bitmap[part_idx], part_off);
+}
+
+/*Converts a net_device to corresponding netdev_index*/
+static u8 get_netdev_index(struct sail *s, struct net_device *dev)
+{
+	u8 i;
+
+	/*checks if the net_device is already seen; if yes then return the
+	 *corresponding index
+	 */
+	for (i = 0; i < s->netdev_count; i++) {
+		if (s->netdevs[i] == dev)
+			return i;
+	}
+	/*If the net_device is not previously seen, then add it to the array*/
+	s->netdevs[s->netdev_count++] = dev;
+	return s->netdev_count - 1;
+}
+
+/* Insert a new chunk to a level at chunk_id-1. Note that Chunk ID
+ * starts from 1, not 0. This function increments cnk_count of sail.
+ * This is why, it is passed by reference
+ */
+static int chunk_insert(u8 *N, u8 *P, struct chunk16 *ck, u32 *cnk_count, u16 chunk_id, u32 cnk_size)
+{
+	long long m;
+
+	if (chunk_id > (*cnk_count + 1) || chunk_id < 1) {
+		pr_err("Invalid chunk_id");
+		return -EINVAL;
+	}
+
+	if (*cnk_count >= NUM_CHUNKS) {
+		pr_err("Cannot insert a new chunk. The chunk array is full");
+		return -EINVAL;
+	}
+
+	/*shift each element one step right to make
+	 *space for the new one
+	 */
+	m = (long long)(*cnk_count) * cnk_size - 1;
+	for (; m >= (chunk_id - 1) * cnk_size; m--) {
+		N[m + cnk_size] = N[m];
+		P[m + cnk_size] = P[m];
+	}
+
+	/*Reset the newly created empty chunk*/
+	m = (chunk_id - 1) * cnk_size;
+	for (; m < chunk_id * cnk_size; m++) {
+		N[m] = 0;
+		P[m] = 0;
+	}
+        
+        if (!ck)
+            goto finish;
+        
+        /*shift each chunk one step right to make
+         *space for the new one
+         */
+        m = (long long)(*cnk_count) - 1;
+        for (; m >= (chunk_id - 1); m--)
+                ck[m + 1] = ck[m];
+
+        /*Reset the newly created empty chunk*/
+        for (m = 0; m < PARTS; m++) {
+                ck[chunk_id - 1].c_bitmap[m] = 0;
+                ck[chunk_id - 1].c_popcnt[m] = 0;
+//                ck[chunk_id - 1].n_bitmap[m] = 0;
+//                ck[chunk_id - 1].n_popcnt[m] = 0;
+        }
+
+finish:
+	++(*cnk_count);
+	return 0;
+}
+
+static int chunk_sm_insert(u8 *N, u8 *P, struct chunk8 *ck, u32 *cnk_count, u16 chunk_id, u32 cnk_size)
+{
+	long long m;
+
+	if (chunk_id > (*cnk_count + 1) || chunk_id < 1) {
+		pr_err("Invalid chunk_id");
+		return -EINVAL;
+	}
+
+	if (*cnk_count >= SM_CHUNKS) {
+		pr_err("Cannot insert a new chunk. The chunk array is full");
+		return -EINVAL;
+	}
+
+	/*shift each element one step right to make
+	 *space for the new one
+	 */
+	m = (long long)(*cnk_count) * cnk_size - 1;
+	for (; m >= (chunk_id - 1) * cnk_size; m--) {
+		N[m + cnk_size] = N[m];
+		P[m + cnk_size] = P[m];
+	}
+
+	/*Reset the newly created empty chunk*/
+	m = (chunk_id - 1) * cnk_size;
+	for (; m < chunk_id * cnk_size; m++) {
+		N[m] = 0;
+		P[m] = 0;
+	}
+        
+        if (!ck)
+            goto finish;
+        
+            /*shift each chunk one step right to make
+             *space for the new one
+             */
+            m = (long long)(*cnk_count) - 1;
+            for (; m >= (chunk_id - 1); m--)
+                    ck[m + 1] = ck[m];
+
+            /*Reset the newly created empty chunk*/
+            for (m = 0; m < SMALL; m++) {
+                    ck[chunk_id - 1].bitmap[m] = 0;
+                    ck[chunk_id - 1].popcnt[m] = 0;
+            }
+        
+
+finish:
+	++(*cnk_count);
+	return 0;
+}
+
+/*Calculate the chunk ID for level 24 based on C16*/
+static u16 calc_ckid24_from_C16(u16 *c16, u32 c16_size, u16 idx16)
+{
+	long long i;
+
+	if (idx16 >= c16_size) {
+		pr_err("Index needs to be smaller than arr_size");
+		return 0;
+	}
+
+	/*Find the first chunk ID to the left and increment that by 1*/
+	for (i = (long long)idx16 - 1; i >= 0; i--) {
+		if (c16[i] > 0)
+			return c16[i] + 1;
+	}
+
+	/*If there is no chunk to the left, then this is the first chunk*/
+	return 1;
+}
+
+/*Update C16 based on the newly inserted chunk*/
+static int update_C16(u16 *c16, u32 c16_size,
+		      u16 idx16, u16 chunk_id)
+{
+	long long i;
+
+	if (idx16 >= c16_size) {
+		pr_err("Invalid index");
+		return -EINVAL;
+	}
+
+	c16[idx16] = chunk_id;
+
+	/* Increment chunk ID to the right */
+	for (i = idx16 + 1; i < c16_size; i++) {
+		if (c16[i] > 0)
+			c16[i]++;
+	}
+
+	return 0;
+}
+
+/*Calculate the chunk ID for upper level based on CK and C of lower level*/
+static u16 calc_ckid(struct chunk16 *ck, u16 ck_idx, u64 ck_off)
+{
+	long long i, j;
+	long long index = 0;
+	u32 part_off, part_idx;
+
+	part_idx = ck_off / 64;
+	part_off = ck_off % 64;
+        
+	/*find the index to C24 where previous chunk ID would be found*/
+	if (ck[ck_idx].c_bitmap[part_idx]) {
+		index = ck[ck_idx].c_popcnt[part_idx] +
+		    POPCNT_OFF(ck[ck_idx].c_bitmap[part_idx], part_off) - 1;
+                if (index >= CNK_16 - 2)
+                    goto err;
+		if (index >= 0)
+			return index + 2;
+	}
+
+	j = (long long)part_idx - 1;
+	for (i = (long long)ck_idx; i >= 0; i--) {
+		for (; j >= 0; j--) {
+			if (ck[i].c_bitmap[j]) {
+				index = ck[i].c_popcnt[j] +
+					POPCNT(ck[i].c_bitmap[j]) - 1;
+                                if (index >= CNK_16 - 2)
+                                    goto err;
+				return index + 2;
+			}
+		}
+		j = PARTS - 1;
+	}
+
+	/*If there is no chunk to the left, then this is the first chunk*/
+	return 1;
+err:
+        pr_err("Invalid index");
+        return -EINVAL;
+}
+
+/*Calculate the chunk ID for upper level based on CK and C of lower level*/
+static u16 calc_ckid_sm(struct chunk8 *ck, u16 ck_idx, u64 ck_off)
+{
+	long long i, j;
+	long long index = 0;
+	u8 part_off, part_idx;
+
+	part_idx = ck_off / 64;
+	part_off = ck_off % 64;
+        
+	/*find the index to C24 where previous chunk ID would be found*/
+	if (ck[ck_idx].bitmap[part_idx]) {
+		index = ck[ck_idx].popcnt[part_idx] +
+		    POPCNT_OFF(ck[ck_idx].bitmap[part_idx], part_off) - 1;
+                if (index >= CNK_16 - 2)
+                    goto err;
+		if (index >= 0)
+			return index + 2;
+	}       
+
+	j = (long long)part_idx - 1;
+	for (i = (long long)ck_idx; i >= 0; i--) {
+		for (; j >= 0; j--) {
+			if (ck[i].bitmap[j]) {
+				index = ck[i].popcnt[j] +
+					POPCNT(ck[i].bitmap[j]) - 1;
+                                if (index >= CNK_16 - 2)
+                                    goto err;
+				return index + 2;
+			}
+		}
+		j = SMALL - 1;
+	}
+        
+	/*If there is no chunk to the left, then this is the first chunk*/
+	return 1;
+err:
+        pr_err("Invalid index");
+        return -EINVAL;
+}
+
+/*Update CK and C based on the newly inserted chunk*/
+static int update_ck_c(struct chunk16 *ck, u32 ck_idx, u32 ck_off, u32 cnk_count)
+{
+	long long i, j;
+	u64 index = 0;
+	u32 part_idx, part_off;
+
+	part_idx = ck_off / 64;
+	part_off = ck_off % 64;
+
+	if (ck[ck_idx].c_bitmap[part_idx] & (1ULL << part_off)) {
+		pr_err("Error: bitmap is already set");
+		return -EINVAL;
+	}
+
+	/*find the index where Chunk ID should be copied to*/
+	if (ck[ck_idx].c_bitmap[part_idx]) {
+		index = ck[ck_idx].c_popcnt[part_idx] +
+			POPCNT_OFF(ck[ck_idx].c_bitmap[part_idx], part_off);
+                if (index >= CNK_16)
+                    goto err;
+		goto index_found;
+	}
+
+	/*Find a chunk to the left which is not empty*/
+	j = (long long)part_idx - 1;
+	for (i = (long long)ck_idx; i >= 0; i--) {
+		for (; j >= 0; j--) {
+			if (ck[i].c_bitmap[j]) {
+				index = ck[i].c_popcnt[j] +
+					POPCNT(ck[i].c_bitmap[j]);
+                                if (index >= CNK_16)
+                                    goto err;
+				goto index_found;
+			}
+		}
+		j = PARTS - 1;
+	}
+
+index_found:
+
+	/*This is the first element of this chunk*/
+	if (!ck[ck_idx].c_bitmap[part_idx])
+		ck[ck_idx].c_popcnt[part_idx] = (u16)index;
+
+	ck[ck_idx].c_bitmap[part_idx] |= (1ULL << part_off);
+
+	/*Update offset of the chunks to the right*/
+	j = part_idx + 1;
+	for (i = (long long)ck_idx; i < cnk_count; i++) {
+		for (; j < PARTS; j++) {
+			if (ck[i].c_bitmap[j])
+				ck[i].c_popcnt[j]++;
+		}
+		j = 0;
+	}
+
+	return 0;
+err:
+        pr_err("Invalid index");
+        return -EINVAL;
+}
+
+/*Update CK and C based on the newly inserted chunk*/
+static int update_ck_sm_c(struct chunk8 *ck, u32 ck_idx, u32 ck_off, u32 cnk_count)
+{
+	long long i, j;
+	u64 index = 0;
+	u8 part_idx, part_off;
+
+	part_idx = ck_off / 64;
+	part_off = ck_off % 64;
+
+	if (ck[ck_idx].bitmap[part_idx] & (1ULL << part_off)) {
+		pr_err("Error: bitmap is already set");
+		return -EINVAL;
+	}
+
+	/*find the index where Chunk ID should be copied to*/
+	if (ck[ck_idx].bitmap[part_idx]) {
+		index = ck[ck_idx].popcnt[part_idx] +
+			POPCNT_OFF(ck[ck_idx].bitmap[part_idx], part_off);
+                if (index >= CNK_16)
+                    goto err;
+                    
+		goto index_found;
+	}
+
+	/*Find a chunk to the left which is not empty*/
+	j = (long long)part_idx - 1;
+	for (i = (long long)ck_idx; i >= 0; i--) {
+		for (; j >= 0; j--) {
+			if (ck[i].bitmap[j]) {
+				index = ck[i].popcnt[j] +
+					POPCNT(ck[i].bitmap[j]);
+                                if (index >= CNK_16)
+                                    goto err;
+				goto index_found;
+			}
+		}
+		j = SMALL - 1;
+	}
+
+index_found:
+
+	/*This is the first element of this chunk*/
+	if (!ck[ck_idx].bitmap[part_idx])
+		ck[ck_idx].popcnt[part_idx] = (u16)index;
+
+	ck[ck_idx].bitmap[part_idx] |= (1ULL << part_off);
+
+	/*Update offset of the chunks to the right*/
+	j = part_idx + 1;
+	for (i = (long long)ck_idx; i < cnk_count; i++) {
+		for (; j < SMALL; j++) {
+			if (ck[i].bitmap[j])
+				ck[i].popcnt[j]++;
+		}
+		j = 0;
+	}
+
+	return 0;
+err:
+        pr_err("Invalid index");
+        return -EINVAL;
+}
+
+int sail_insert(struct sail *s, u64 key, u8 prefix_len, struct net_device *dev)
+{
+	int i, j;
+	u8 *n16, *p16, *n32, *p32, *n40, *p40, *n48, *p48, *n64, *p64;
+	u16 *c16;
+	struct chunk16 *ck32;
+        struct chunk8 *ck40, *ck48;
+	u16 chunk_id = 0;
+        u32 n_idx;
+        u16 ck_idx;
+        u64 ck_off;
+	u16 n16_idx;/*Index to N16, P16 and C16*/	     
+	u32 part_idx, part_off;
+	u32 num_leafs;/*Number of leafs need to be inserted for this prefix*/
+	u8 netdev_index = get_netdev_index(s, dev);
+	int err = 0;
+        u64 bitmap;
+        u64 tmp;
+        u32 new_prefixes = 0;
+
+	spin_lock(&s->lock);
+
+	/* Default route */
+	if (prefix_len == 0) {
+		s->def_nh = netdev_index;
+		goto finish;
+	}
+
+	/* Preallocate all the arrays at once*/
+	if (!s->N16) {
+		n16 = kcalloc(LEVEL16_SIZE, sizeof(*n16), GFP_ATOMIC);
+		p16 = kcalloc(LEVEL16_SIZE, sizeof(*p16), GFP_ATOMIC);
+		c16 = kcalloc(LEVEL16_SIZE, sizeof(*c16), GFP_ATOMIC);
+                
+                n32 = kcalloc(ARR_SIZE, sizeof(*n32), GFP_ATOMIC);
+                p32 = kcalloc(ARR_SIZE, sizeof(*p32), GFP_ATOMIC);
+                ck32 = kcalloc(NUM_CHUNKS, sizeof(*ck32), GFP_ATOMIC);
+
+                n40 = kcalloc(ARR_SIZE_SM, sizeof(*n40), GFP_ATOMIC);
+                p40 = kcalloc(ARR_SIZE_SM, sizeof(*p40), GFP_ATOMIC);
+                ck40 = kcalloc(SM_CHUNKS, sizeof(*ck40), GFP_ATOMIC);
+                
+                n48 = kcalloc(ARR_SIZE_SM, sizeof(*n48), GFP_ATOMIC);
+                p48 = kcalloc(ARR_SIZE_SM, sizeof(*p48), GFP_ATOMIC);
+                ck48 = kcalloc(SM_CHUNKS, sizeof(*ck48), GFP_ATOMIC);
+
+                n64 = kcalloc(ARR_SIZE, sizeof(*n64), GFP_ATOMIC);
+                p64 = kcalloc(ARR_SIZE, sizeof(*p64), GFP_ATOMIC);
+                
+		if (!n16 || !c16 || !p16 || !n32 || !p32 || !ck32 ||
+                        !n40 || !p40 || !ck40 || 
+                        !n48 || !p48 || !ck48 || !n64 || !p64) {
+			kfree(n16);
+			kfree(c16);
+			kfree(p16);
+                        kfree(n32);
+                        kfree(p32);
+                        kfree(ck32);
+                        kfree(n40);
+                        kfree(p40);
+                        kfree(ck40);
+                        kfree(n48);
+                        kfree(p48);
+                        kfree(ck48);
+                        kfree(n64);
+                        kfree(p64);
+			pr_err("Out of memory while preallocating  SAIL");
+			goto error;
+		}
+
+		RCU_INIT_POINTER(s->N16, n16);
+		RCU_INIT_POINTER(s->P16, p16);
+		RCU_INIT_POINTER(s->C16, c16);
+
+                RCU_INIT_POINTER(s->N32, n32);
+                RCU_INIT_POINTER(s->P32, p32);
+                RCU_INIT_POINTER(s->CK32, ck32);
+                s->n32_cnt = 0;
+
+                RCU_INIT_POINTER(s->N40, n40);
+                RCU_INIT_POINTER(s->P40, p40);
+                RCU_INIT_POINTER(s->CK40, ck40);
+                
+                RCU_INIT_POINTER(s->N48, n48);
+                RCU_INIT_POINTER(s->P48, p48);
+                RCU_INIT_POINTER(s->CK48, ck48);
+
+                RCU_INIT_POINTER(s->N64, n64);
+                RCU_INIT_POINTER(s->P64, p64); 
+                
+//		synchronize_rcu();
+	}
+
+	/*Eextract 16 bits from MSB.*/
+	n16_idx = key >> 48;
+
+	if (prefix_len <= 16) {
+		/*All the leafs in level 1~16 will be stored in level 16.*/
+		num_leafs = 1U << (16 - prefix_len);
+		for (i = 0; i < num_leafs; i++) {
+			/*Longer prefix exists*/
+			if (s->P16[n16_idx + i] > prefix_len)
+				continue;
+			s->N16[n16_idx + i] = netdev_index;
+			s->P16[n16_idx + i] = prefix_len;
+		}
+		goto finish;
+	}
+
+	/* The length of the prefix is 21~64. So need to check if there is a
+	 * chunk for this prefix in level 24. C16[..] = 0 indicates that there
+	 * is no chunk, so need to insert one. The insertion works as
+	 * following:
+	 * 1. Calculate the chunk ID to level 24 from C16. The Chunk ID
+	 * indicates where the new chunk should be inserted.
+	 * 2. Insert a new chunk in level 24
+	 * 3. Update the C16[] based on the newly inserted chunk
+	 */
+	if (s->C16[n16_idx] == 0) {
+		/*Step 1*/
+		chunk_id = calc_ckid24_from_C16(s->C16, LEVEL16_SIZE, n16_idx);
+		if (!chunk_id)
+			goto error;
+		/*Step 2*/
+		err = chunk_insert(s->N32, s->P32, s->CK32, &s->cnk32_count, chunk_id, CNK_16);
+		if (err) {
+                        pr_err("Could not insert chunk to level 32");
+			goto error;
+                }
+		/*Step 3*/
+		err = update_C16(s->C16, LEVEL16_SIZE, n16_idx, chunk_id);
+		if (err)
+			goto error;
+	}
+
+	/*Extract bit 17~32 from MSB and calculate index to level 32*/
+//	n_idx = (s->C16[n16_idx] - 1) * CNK_16 + ((key >> 32) & 0XFFFF);
+	ck_idx = s->C16[n16_idx] - 1;
+	ck_off = (key >> 32) & 0XFFFF;
+        part_idx = ck_off / 64;
+        part_off = ck_off % 64;
+        
+	if (prefix_len <= 32) {
+		/*All the leafs in level 17~32 will be stored in level 32.*/
+		num_leafs = 1U << (32 - prefix_len);
+		for (i = 0, tmp = 1ULL << part_off; i < num_leafs; i++, tmp <<= 1) {
+                    n_idx = calc_n_idx(&(s->CK32[ck_idx]), ck_off + i);
+                    if (s->CK32[ck_idx].n_bitmap[part_idx] & tmp) {                        
+			/*Longer prefix exists*/
+                        if (s->P32[n_idx] > prefix_len)
+                                continue;
+                        s->N32[n_idx] = netdev_index;
+                        s->P32[n_idx] = prefix_len;
+                    } else {
+                        if (n_idx >= ARR_SIZE)
+                            goto error;
+                        /*FIXME: If this loop goes for long, that breaks the kerenel*/
+                        for (j = (int)s->n32_cnt - 1; j >= (int)n_idx; j--) {
+                                s->N32[j + 1] = s->N32[j];
+                                s->P32[j + 1] = s->P32[j];
+                        }
+                        s->N32[n_idx] = netdev_index;
+                        s->P32[n_idx] = prefix_len;
+                        new_prefixes++;
+                        s->n32_cnt++;
+                    }
+		}
+                j = part_idx + 1;
+                for (i = ck_idx; i < NUM_CHUNKS; i++) {
+                    for (; j < PARTS; j++) {
+                        s->CK32[i].n_popcnt[j] += new_prefixes;
+                    }
+                    j = 0;
+                }
+                bitmap = ((1ULL << num_leafs) - 1) << part_off;
+                s->CK32[ck_idx].n_bitmap[part_idx] |= bitmap;
+		goto finish;
+	}
+
+
+        /* The length of the prefix is 33-64, but there is no chunk for
+         * this prefix in the next level. So do the same as before
+         */
+
+        if (!(s->CK32[ck_idx].c_bitmap[part_idx] & (1ULL << part_off))) {
+                chunk_id = calc_ckid(s->CK32, ck_idx, ck_off);
+                if (!chunk_id)
+                        goto error;
+                err = chunk_sm_insert(s->N40, s->P40, s->CK40, &s->cnk40_count, chunk_id, CNK_8);
+		if (err) {
+                        pr_err("Could not insert chunk to level 40");
+			goto error;
+                }
+                err = update_ck_c(s->CK32, ck_idx, ck_off, s->cnk32_count);
+                if (err)
+                        goto error;
+        }
+        ck_idx = calc_ck_idx(&(s->CK32[ck_idx]), ck_off);
+        ck_off = (key >> 24) & 0XFF;
+        /*Extract 16 more bits and calculate index to level 48, 64,..*/
+        n_idx = ck_idx * CNK_8 + ck_off;
+
+        if (prefix_len <= 40) {
+                /*All the leafs in level 33~48 will be stored in level 48.*/
+                num_leafs = 1U << (40 - prefix_len);
+                for (i = 0; i < num_leafs; i++) {
+                        /*Longer prefix exists*/
+                        if (s->P40[n_idx + i] > prefix_len)
+                                continue;
+                        s->N40[n_idx + i] = netdev_index;
+                        s->P40[n_idx + i] = prefix_len;
+                }
+                goto finish;
+        }
+        
+        /* The length of the prefix is 33-64, but there is no chunk for
+         * this prefix in the next level. So do the same as before
+         */
+        part_idx = ck_off / 64;
+        part_off = ck_off % 64;
+        if (!(s->CK40[ck_idx].bitmap[part_idx] & (1ULL << part_off))) {
+                chunk_id = calc_ckid_sm(s->CK40, ck_idx, ck_off);
+                if (!chunk_id)
+                        goto error;
+                err = chunk_sm_insert(s->N48, s->P48, s->CK48, &s->cnk48_count, chunk_id, CNK_8);
+		if (err) {
+                        pr_err("Could not insert chunk to level 48");
+			goto error;
+                }
+                err = update_ck_sm_c(s->CK40, ck_idx, ck_off, s->cnk40_count);
+                if (err)
+                        goto error;
+        }
+        
+        ck_idx = calc_ck_idx_sm(&(s->CK40[ck_idx]), ck_off);
+        ck_off = (key >> 16) & 0XFF;
+        /*Extract 16 more bits and calculate index to level 48, 64,..*/
+        n_idx = ck_idx * CNK_8 + ck_off;        
+
+        if (prefix_len <= 48) {
+                /*All the leafs in level 33~48 will be stored in level 48.*/
+                num_leafs = 1U << (48 - prefix_len);
+                for (i = 0; i < num_leafs; i++) {
+                        /*Longer prefix exists*/
+                        if (s->P48[n_idx + i] > prefix_len)
+                                continue;
+                        s->N48[n_idx + i] = netdev_index;
+                        s->P48[n_idx + i] = prefix_len;
+                }
+                goto finish;
+        }
+        
+        /* The length of the prefix is 49-64, but there is no chunk for
+         * this prefix in the next level. So do the same as before
+         */
+        part_idx = ck_off / 64;
+        part_off = ck_off % 64;
+        if (!(s->CK48[ck_idx].bitmap[part_idx] & (1ULL << part_off))) {
+                chunk_id = calc_ckid_sm(s->CK48, ck_idx, ck_off);
+                if (!chunk_id)
+                        goto error;
+                err = chunk_insert(s->N64, s->P64, NULL, &s->cnk64_count, chunk_id, CNK_16);
+		if (err) {
+                        pr_err("Could not insert chunk to level 64");
+			goto error;
+                }
+                err = update_ck_sm_c(s->CK48, ck_idx, ck_off, s->cnk48_count);
+                if (err)
+                        goto error;
+        }
+
+        /*Extract 16 more bits and calculate index to level 48, 64,..*/
+        ck_idx = calc_ck_idx_sm(&(s->CK48[ck_idx]), ck_off);
+        ck_off = key & 0XFFFF;
+        n_idx = ck_idx * CNK_16 + ck_off;
+
+        if (prefix_len <= 64) {
+                /* All the leafs in level 49~64 will be stored in level 64.*/
+                num_leafs = 1U << (64 - prefix_len);
+                for (i = 0; i < num_leafs; i++) {
+                        /*Longer prefix exists*/
+                        if (s->P64[n_idx + i] > prefix_len)
+                                continue;
+                        s->N64[n_idx + i] = netdev_index;
+                        s->P64[n_idx + i] = prefix_len;
+                }
+                goto finish;
+        }       
+
+error:
+	pr_err("Something went wrong in route insertion");
+finish:
+	spin_unlock(&s->lock);
+	return err;
+}
+
+int sail_lookup(const struct sail *s, const u64 key, struct net_device **dev)
+{
+	u8 *n16, *n;
+	u16 *c16;
+	struct chunk16 *ck;        
+        struct chunk8 *ck_sm; /*Small chunk*/
+	u8 netdev_index = s->def_nh;
+	u16 n16_idx;/*Index to N16 and C16*/
+	u32 n_idx;/*Index to N and C*/
+	u32 ck_idx;/*Index to CK*/
+	u64 ck_off;/*Index to CK*/
+	u32 part_idx, part_off;
+
+	rcu_read_lock();
+
+	/*extract 16 bits from LSB*/
+	n16_idx = key >> 48;
+	n16 = rcu_dereference(s->N16);
+	if (unlikely(!n16))
+		goto finish;
+
+	if (likely(n16[n16_idx] != 0))
+		netdev_index = n16[n16_idx];
+
+	/*Check if there is a longer prefix; if yes, extract bit 17~32
+	 *  and calculate index to N32
+	 */
+	c16 = rcu_dereference(s->C16);
+	if (likely(c16[n16_idx] != 0)) {
+		ck_idx = c16[n16_idx] - 1;
+		ck_off = (key >> 32) & 0XFFFF;
+		n_idx = ck_idx * CNK_16 + ck_off;
+	} else {
+		goto finish;
+	}
+
+        /*Check if there is a longer prefix; if yes, extract more 16 bits
+         * from MSB and calculate index to N48
+         */
+        ck = rcu_dereference(s->CK32);
+        part_idx = ck_off / 64;
+        part_off = ck_off % 64;
+
+        /*Find corresponding next-hop in level 32*/
+        if (s->CK32[ck_idx].n_bitmap[part_idx] & (1ULL << part_off)) {
+            n = rcu_dereference(s->N32);
+            n_idx = calc_n_idx(&(s->CK32[ck_idx]), ck_off);
+            netdev_index = n[n_idx];      
+        }
+
+        if (likely(ck[ck_idx].c_bitmap[part_idx] & (1ULL << part_off))) {
+                ck_idx = ck[ck_idx].c_popcnt[part_idx] +
+                        POPCNT_OFF(ck[ck_idx].c_bitmap[part_idx], part_off);
+                ck_off = (key >> 24) & 0XFF;
+                n_idx = ck_idx * CNK_8 + ck_off;
+        } else {
+                goto finish;
+        }
+
+        n = rcu_dereference(s->N40);
+        if (likely(n[n_idx] != 0))
+                netdev_index = n[n_idx];
+        
+        /*Check if there is a longer prefix; if yes, extract more 16 bits
+         * from MSB and calculate index to N48
+         */
+        ck_sm = rcu_dereference(s->CK40);
+        part_idx = ck_off / 64;
+        part_off = ck_off % 64;
+        if (likely(ck_sm[ck_idx].bitmap[part_idx] & (1ULL << part_off))) {
+                ck_idx = ck_sm[ck_idx].popcnt[part_idx] +
+                        POPCNT_OFF(ck_sm[ck_idx].bitmap[part_idx], part_off);
+                ck_off = (key >> 16) & 0XFF;
+                n_idx = ck_idx * CNK_8 + ck_off;
+        } else {
+                goto finish;
+        }
+
+        n = rcu_dereference(s->N48);
+        if (likely(n[n_idx] != 0))
+                netdev_index = n[n_idx];
+        
+        /*Check if there is a longer prefix; if yes, extract more 16 bits
+         * from MSB and calculate index to N64
+         */
+        ck_sm = rcu_dereference(s->CK48);
+        part_idx = ck_off / 64;
+        part_off = ck_off % 64;
+        if (likely(ck_sm[ck_idx].bitmap[part_idx] & (1ULL << part_off))) {
+                ck_idx = ck_sm[ck_idx].popcnt[part_idx] +
+                        POPCNT_OFF(ck_sm[ck_idx].bitmap[part_idx], part_off);
+                ck_off = key & 0XFFFF;
+                n_idx = ck_idx * CNK_16 + ck_off;
+        } else {
+                goto finish;
+        }
+
+        n = rcu_dereference(s->N64);
+        if (likely(n[n_idx] != 0))
+                netdev_index = n[n_idx];
+
+finish:
+	*dev = s->netdevs[netdev_index];
+	rcu_read_unlock();
+	return 0;
+}
diff --git a/net/ipv6/route.c b/net/ipv6/route.c
index 826b14d..dc6f893 100644
--- a/net/ipv6/route.c
+++ b/net/ipv6/route.c
@@ -678,18 +678,20 @@ static struct fib6_info *find_match(struct fib6_info *rt, int oif, int strict,
 		match_do_rr = true;
 		m = 0; /* lowest valid score */
 	} else if (m == RT6_NUD_FAIL_HARD) {
-		goto out;
+//Tamim, undo
+//		goto out;
 	}
 
 	if (strict & RT6_LOOKUP_F_REACHABLE)
 		rt6_probe(rt);
 
 	/* note that m can be RT6_NUD_FAIL_PROBE at this point */
-	if (m > *mpri) {
+//Tamim, undo
+//	if (m > *mpri) {
 		*do_rr = match_do_rr;
 		*mpri = m;
 		match = rt;
-	}
+//	}
 out:
 	return match;
 }
@@ -1146,6 +1148,32 @@ struct rt6_info *rt6_lookup(struct net *net, const struct in6_addr *daddr,
 }
 EXPORT_SYMBOL(rt6_lookup);
 
+#if IS_ENABLED(CONFIG_FIB_SAIL_XDP)
+static void get_addr (struct in6_addr *addr, u64 key)
+{
+    __be16 *addr_ipv6 = addr->in6_u.u6_addr16;
+    addr_ipv6[0] = htons(key >> 48);
+    addr_ipv6[1] = htons((key >> 32) & 0XFFFF);
+    addr_ipv6[2] = htons((key >> 16) & 0XFFFF);
+    addr_ipv6[3] = htons(key & 0XFFFF);
+    addr_ipv6[4] = 0;
+    addr_ipv6[5] = 0;
+    addr_ipv6[6] = 0;
+    addr_ipv6[7] = 0;
+}
+
+static u64 get_key (struct in6_addr addr)
+{
+    u64 key;
+    __be16 *addr_ipv6 = addr.in6_u.u6_addr16;
+    key = ((__force u64)(ntohs(addr_ipv6[0])) << 48) |
+          ((__force u64)(ntohs(addr_ipv6[1])) << 32) |
+          ((__force u64)(ntohs(addr_ipv6[2])) << 16) |
+          (__force u64)(ntohs(addr_ipv6[3]));
+    return key;
+}
+#endif
+
 /* ip6_ins_rt is called with FREE table->tb6_lock.
  * It takes new route entry, the addition fails by any reason the
  * route is released.
@@ -1157,7 +1185,13 @@ static int __ip6_ins_rt(struct fib6_info *rt, struct nl_info *info,
 {
 	int err;
 	struct fib6_table *table;
+#if IS_ENABLED(CONFIG_FIB_SAIL_XDP)
 
+        if (rt->fib6_dst.plen <= 64) {
+//            pr_err("Inserting %pI6c %d => %s", &rt->fib6_dst.addr, rt->fib6_dst.plen, rt->fib6_nh.nh_dev->name);
+            sail_insert(&rt->fib6_table->sail, get_key(rt->fib6_dst.addr), rt->fib6_dst.plen, rt->fib6_nh.nh_dev);            
+        }
+#endif
 	table = rt->fib6_table;
 	spin_lock_bh(&table->tb6_lock);
 	err = fib6_add(&table->tb6_root, rt, info, extack);
@@ -1854,6 +1888,14 @@ struct fib6_info *fib6_table_lookup(struct net *net, struct fib6_table *table,
 	return f6i;
 }
 
+#if IS_ENABLED(CONFIG_FIB_SAIL_XDP)
+static u64 global_key = 0X2001020000000000;
+#define NUM_TEST_CASES 1000
+static u64 arr [NUM_TEST_CASES];
+static u64 increment = 0X100000000;
+static spinlock_t lock_gen;
+#endif
+
 struct rt6_info *ip6_pol_route(struct net *net, struct fib6_table *table,
 			       int oif, struct flowi6 *fl6,
 			       const struct sk_buff *skb, int flags)
@@ -1861,6 +1903,10 @@ struct rt6_info *ip6_pol_route(struct net *net, struct fib6_table *table,
 	struct fib6_info *f6i;
 	struct rt6_info *rt;
 	int strict = 0;
+#if IS_ENABLED(CONFIG_FIB_SAIL_XDP)
+        struct net_device *dev;
+        int i;
+#endif
 
 	strict |= flags & RT6_LOOKUP_F_IFACE;
 	strict |= flags & RT6_LOOKUP_F_IGNORE_LINKSTATE;
@@ -1869,6 +1915,46 @@ struct rt6_info *ip6_pol_route(struct net *net, struct fib6_table *table,
 
 	rcu_read_lock();
 
+#if IS_ENABLED(CONFIG_FIB_SAIL_XDP)
+        
+//        spin_lock(&lock_gen);
+        
+        for (i = 0; i < NUM_TEST_CASES; i++) {
+            arr[i] = global_key;
+            if ( global_key >= 0X2001800000000000) {
+                global_key = 0X2001020000000000;
+                if (increment == 0X100000000)
+                    increment = 0X10000;
+                else if (increment == 0X10000)
+                    increment = 0X1000;
+                else 
+                    increment = 0X100000000;
+            } else { 
+                global_key += increment;
+            }
+        }                
+
+//        for (i = 0; i < NUM_TEST_CASES; i++) {
+//            get_addr(&fl6->daddr, arr[i]);
+//            pr_err("arr[%d] = %pI6c", i, &fl6->daddr);
+//        }
+        
+        for (i = 0; i < NUM_TEST_CASES; i++) {
+            sail_lookup(&net->ipv6.fib6_main_tbl->sail, arr[i], &dev);
+            get_addr(&fl6->daddr, arr[i]);            
+            f6i = fib6_table_lookup(net, net->ipv6.fib6_main_tbl, oif, fl6, strict);
+            if (f6i->fib6_nh.nh_dev->name && dev->name &&
+                    strcmp(f6i->fib6_nh.nh_dev->name, dev->name) &&
+                    strcmp(f6i->fib6_nh.nh_dev->name, "wlo1") &&
+                    strcmp(f6i->fib6_nh.nh_dev->name, "lo")) {
+                pr_err("%d.Trie: %pI6c => %s", i, &fl6->daddr, f6i->fib6_nh.nh_dev->name);
+                pr_err("%d.SAIL: %pI6c => %s", i, &fl6->daddr, dev->name);
+            }
+        }
+        
+//        spin_unlock(&lock_gen);
+
+#endif
 	f6i = fib6_table_lookup(net, table, oif, fl6, strict);
 	if (f6i->fib6_nsiblings)
 		f6i = fib6_multipath_select(net, f6i, fl6, oif, skb, strict);
@@ -3179,7 +3265,14 @@ static int __ip6_del_rt(struct fib6_info *rt, struct nl_info *info)
 	struct net *net = info->nl_net;
 	struct fib6_table *table;
 	int err;
-
+#if IS_ENABLED(CONFIG_FIB_SAIL_XDP)
+        if (rt->fib6_dst.plen <= 64) {
+//            pr_err("Deleting %pI6c/%d", &rt->fib6_dst.addr, rt->fib6_dst.plen);
+//            sail_delete(&rt->fib6_table->sail, get_key(rt->fib6_dst.addr), rt->fib6_dst.plen);
+        }
+#endif
+	table = rt->fib6_table;
+        
 	if (rt == net->ipv6.fib6_null_entry) {
 		err = -ENOENT;
 		goto out;
